src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -371,146 +371,154 @@ internal unsafe void Start()
                 {
                     // Flow suppressed, just run here
-                    ProcessEvents(numEvents.ToInt32(), eventPaths, eventFlags, eventIds, watcher);
+                    ProcessEvents(numEvents.ToInt32(), eventPaths, new Span<FSEventStreamEventFlags>(eventFlags, numEvents.ToInt32()), new Span<FSEventStreamEventId>(eventIds, numEvents.ToInt32()), watcher);
                 }
                 else
                 {
                     ExecutionContext.Run(
                         context,
-                        (object o) => ((RunningInstance)o).ProcessEvents(numEvents.ToInt32(), eventPaths, eventFlags, eventIds, watcher),
+                        (object o) => ((RunningInstance)o).ProcessEvents(numEvents.ToInt32(), eventPaths, new Span<FSEventStreamEventFlags>(eventFlags, numEvents.ToInt32()), new Span<FSEventStreamEventId>(eventIds, numEvents.ToInt32()), watcher),
                         this);
                 }
             }
 
             private unsafe void ProcessEvents(int numEvents,
                 byte** eventPaths,
-                FSEventStreamEventFlags* eventFlags,
-                FSEventStreamEventId* eventIds,
+                Span<FSEventStreamEventFlags> eventFlags,
+                Span<FSEventStreamEventId> eventIds,
                 FileSystemWatcher watcher)
             {
-                // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll add it to this
-                // list so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
-                List<FSEventStreamEventId> handledRenameEvents = null;
-                Memory<char>[] events = new Memory<char>[numEvents];
-                ParseEvents();
+                // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll store the event id
+                // so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
+                int? handledRenameEvents = null;
 
                 for (int i = 0; i < numEvents; i++)
                 {
-                    ReadOnlySpan<char> path = events[i].Span;
-                    Debug.Assert(path[path.Length - 1] != '/', "Trailing slashes on events is not supported");
-
-                    // Match Windows and don't notify us about changes to the Root folder
-                    if (_fullDirectory.Length >= path.Length && path.Equals(_fullDirectory.AsSpan(0, path.Length), StringComparison.OrdinalIgnoreCase))
+                    using (var parsedEvent = ParseEvent(eventPaths[i]))
                     {
-                        continue;
-                    }
+                        var path = parsedEvent.Path;
+                        Debug.Assert(path[path.Length - 1] != '/', "Trailing slashes on events is not supported");
```

Comment:
Nit: as long as you're editing this, if you want to be fancy this can now be `path[^1]` instead of `path[path.Length - 1]`